internal/controller/management/upgrade/freight.go

```diff
@@ -0,0 +1,111 @@
+package upgrade
+
+import (
+	"context"
+
+	log "github.com/sirupsen/logrus"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+
+	kargoapi "github.com/akuity/kargo/api/v1alpha1"
+	"github.com/akuity/kargo/internal/logging"
+)
+
+// freightReconciler reconciles Freight resources to upgrade them from
+// v0.4.0-compatible to v0.5.0-compatible.
+type freightReconciler struct {
+	client client.Client
+}
+
+// SetupFreightReconcilerWithManager initializes a freightReconciler and
+// registers it with the provided Manager.
+func SetupFreightReconcilerWithManager(mgr manager.Manager) error {
+	notV050CompatiblePredicate, err := getNotV050CompatiblePredicate()
+	if err != nil {
+		return err
+	}
+	_, err = ctrl.NewControllerManagedBy(mgr).
+		For(&kargoapi.Freight{}).
+		WithEventFilter(ignoreDeletesPredicate()).
+		WithEventFilter(notV050CompatiblePredicate).
+		Build(&freightReconciler{
+			client: mgr.GetClient(),
+		})
+	return err
+}
+
+// Reconcile is part of the main Kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+func (f *freightReconciler) Reconcile(
+	ctx context.Context,
+	req ctrl.Request,
+) (ctrl.Result, error) {
+	logger := logging.LoggerFromContext(ctx).WithFields(log.Fields{
+		"namespace": req.NamespacedName.Namespace,
+		"freight":   req.NamespacedName.Name,
+	})
+	logger.Debug("reconciling Freight")
+
+	// Find the Freight
+	freight, err := kargoapi.GetFreight(ctx, f.client, req.NamespacedName)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	if freight == nil {
+		// Ignore if not found. This can happen if the Freight was deleted after the
+		// current reconciliation request was issued.
+		return ctrl.Result{}, nil // Do not requeue
+	}
+
+	// None of these things should really happen
+	if len(freight.OwnerReferences) == 0 {
+		logger.Warning("skipping Freight with no OwnerReferences")
```

Comment:
If you have existing (or new?) Freight created by a `v0.5.0` release (or current `main`), this will actually (continuously) happen each time the management controller is started:

```
time="2024-03-28T12:03:43Z" level=warning msg="skipping Freight with no OwnerReferences" freight=319c6e315e081836c1ef597b08cc22811c84af50 namespace=kargo-advanced
time="2024-03-28T12:03:43Z" level=debug msg="reconciling Freight" freight=7baad20879af2b1c2e1c039120d52ca60cdf0218 namespace=kargo-advanced
time="2024-03-28T12:03:43Z" level=warning msg="skipping Freight with no OwnerReferences" freight=7baad20879af2b1c2e1c039120d52ca60cdf0218 namespace=kargo-advanced
```